Breadcrumbs

Twixes opened the pull request:
## Changes

🍞crumbs. ~~WIP.~~

Twixes committed:
Add new sidebar base with project indicator

Twixes committed:
Point TopBar logo to app default path

Twixes committed:
Fix sidebar sizing

Twixes committed:
Make sidebar toggleable

Twixes committed:
Constrain Popup max dimensions for mobile

Twixes committed:
Fix `.main-app-content` width

Twixes committed:
Put icons story at top level of Storybook

Twixes committed:
Add pages to SideBar

Twixes committed:
Update HelpButton for Lemon Design

Twixes committed:
Use fresh icons

Twixes committed:
Add breadcrumbs base

Twixes committed:
Fix some edge cases

Twixes committed:
Merge branch 'master' into breadcrumbs

Twixes committed:
Flesh breadcrumbs out

Twixes committed:
Merge branch 'master' into breadcrumbs

Twixes committed:
Fix typing

Twixes commented:
So, help yourself to some crumbs. Try this out locally and let me know what you think about the experience!
Here are some previews:
- Organization-based:
![Zrzut ekranu 2021-11-8 o 19 06 27](https://user-images.githubusercontent.com/4550621/140794754-1e66097e-04a6-4364-b09f-444e8f20ddc8.png)
- Project-based, level 0:
![Zrzut ekranu 2021-11-8 o 19 02 47](https://user-images.githubusercontent.com/4550621/140794406-a242700c-f6b4-425f-af51-660b8c284445.png)
- Project-based, level 1:
![Zrzut ekranu 2021-11-8 o 18 45 29](https://user-images.githubusercontent.com/4550621/140794408-4d452717-f548-45bd-b2e8-a665fe99e03f.png)
- Instance-level (system status or licenses):
![Zrzut ekranu 2021-11-8 o 18 37 28](https://user-images.githubusercontent.com/4550621/140794412-743bbf45-bf2c-4bb2-ade8-591998df001f.png)
- Personal (i.e. account settings):
![Zrzut ekranu 2021-11-8 o 18 36 24](https://user-images.githubusercontent.com/4550621/140794414-0e825157-ceed-4286-a6db-f7497051e9ca.png)

Twixes committed:
Fix identifiers

paolodamico commented on frontend/src/scenes/sceneTypes.ts:
```diff
@@ -87,6 +87,12 @@ export interface SceneConfig {
     /** Hides demo project warnings (DemoWarning.tsx) */
     hideDemoWarnings?: boolean
-    /** Route requires project access */
+    /** Account management */
```
```suggestion
    /** Part of personal account management - Required for breadcrumbs  */
```

paolodamico committed:
Merge branch 'master' into breadcrumbs

paolodamico commented on frontend/src/layout/lemonade/SideBar/index.scss:
```diff
@@ -0,0 +1,53 @@
+@import '~/vars';
```
I think these files (`frontend/src/layout/lemonade/SideBar/index.tsx` & `frontend/src/layout/lemonade/SideBar/index.scss`) are a mistake, we already have them in master as `frontend/src/layout/lemonade/SideBar/SideBar.tsx`), and those seem to work correctly. Didn't remove them myself in case you made a change that should be included.

paolodamico reviewed (changes requested):
Looking great, did an initial pass at this. Some comments inline. In addition, some UX issues:

1. I strongly recommend adding the current view as part of the breadcrumbs (something like this), otherwise it's hard to know exactly where you are.
<img width="401" alt="" src="https://user-images.githubusercontent.com/5864173/140836024-926074a8-0323-4c5d-a4f4-81c5ef845dc0.png">

2. It was confusing that some breadcrumbs are underlined and others not (see comment 3 too). In addition, "Insights" label is particularly confusing in this context, I wonder if for breadcrumbs we should call this "Insight list". Thoughts @clarkus ?
<img width="336" alt="" src="https://user-images.githubusercontent.com/5864173/140836116-5609657a-a94e-492a-9150-212139c45bd6.png">

3. Intuitively I would expect to be able to switch the project/org from here, if I'm seeing the name and context here.
<img width="299" alt="" src="https://user-images.githubusercontent.com/5864173/140836158-cc014c8b-6534-4961-a3b4-45e11f3dfc36.png">

4. Breadcrumbs disappear on actions. Further, I would probably expect to see Actions as a separate thing, i.e. Hogflix > Demo app > Events > Actions.
<img width="509" alt="" src="https://user-images.githubusercontent.com/5864173/140836222-feb83441-b37c-4a42-84c9-f997fd71a343.png">

clarkus commented:
> 1. I strongly recommend adding the current view as part of the breadcrumbs (something like this), otherwise it's hard to know exactly where you are.

The header is part of this pattern. It's meant to represent the current item. I do not think we should repeat that in the breadcrumbs. 

> 1. It was confusing that some breadcrumbs are underlined and others not (see comment 3 too). In addition, "Insights" label is particularly confusing in this context, I wonder if for breadcrumbs we should call this "Insight list". Thoughts @clarkus ?

I think these should match the names in the navigation. I'd rather go with "Insights" and help users learn what that implies. Also once you click it, you can easily confirm it's a list of insights. 

> 1. Intuitively I would expect to be able to switch the project/org from here, if I'm seeing the name and context here.

Not part of this by design for now. It might be good to add this in the future if it turns out to be a problem for users (I already have designs for it), just don't want to overload this component for now. The main job of the breadcrumbs is to give you a better sense of current place, not so much to offer parity with the collapsed navigation. 

> 1. Breadcrumbs disappear on actions. Further, I would probably expect to see Actions as a separate thing, i.e. Hogflix > Demo app > Events > Actions.

I have an updated page design for this feature at https://www.figma.com/file/9yWtngNb1AIuf6KmXaEPJA/App-doodles?node-id=1121%3A143574. I think we'd just show the org and project in the breadcrumbs here (much like we do for insights). The header would set the context of Events & Actions, and the tabs would then take over setting the specific feature section. 

One thing we should change is the placement of that demo banner. That's mean to be page level, so ideally it'd be shown above the page header or below it. Header here would apply to anything from the breadcrumbs down to the horizontal rule used to delimit the page content from the header.

mariusandra commented:
1. Some issues with responsiveness
![image](https://user-images.githubusercontent.com/53387/140898365-cb26c603-bd80-4e60-a174-a8febdcb9138.png)

2. Personally, I'm in the same boat as @paolodamico and it feels really weird that the breadcrumbs don't include the current URL. I believe it impedes usability... and this is also the approach [recommended by my go to source for UX guidelines](https://www.nngroup.com/articles/breadcrumbs/) (NN Group). Point nr 4: **Include the current page as the last item in the breadcrumb trail.**

I would even go so far to say that I would prefer if we remove the page title, and put it in the breadcrumbs... versus not having the page title in the breadcrumbs. We're duplicating information here anyway, so what's one extra link. Plus it really makes navigation nicer.

3. Going Dash -> Insight makes the crumbs change in a weird way
![2021-11-09 10 31 25](https://user-images.githubusercontent.com/53387/140898772-14659448-a456-43f1-a34f-db790be1bad7.gif)

(made worse by the fact that the current page is not in the breadcrumbs)

mariusandra reviewed (changes requested):
Officially requesting changes as well, to prevent the github bot from reminding me of this.

Twixes committed:
Address feedback

Twixes committed:
Limit `SceneConfig` additions

Twixes committed:
Merge branch 'master' into breadcrumbs

Twixes committed:
`InstanceLicenses` → `Licenses`

Twixes commented:
Thank yall for the feedback!
I've acted upon it in the following way:
1. Added current location as the final breadcrumb. It shows the current item's name/key, except for the insight view, which doesn't have a solid way of getting the currently edited item from outside the page (`fromItem` crap which is due for being significantly refactored anyway for cleaner URLs).
2. Removed underlines from links. Though now it's a bit less clear which breadcrumb items are links.
3. Did NOT add org/project switching. It would require expanding the scope significantly, while also introducing complexities like where to redirect when switching on a specific item's page (i.e. feature flag `foo` may exist in a project, but likely doesn't in another one).
4. Made sure breadcrumbs work on all Events & actions tabs.
5. Improved responsiveness by making the breadcrumbs x-scrollable instead of having them shrink awkwardly.

As for the Dashboards → Insights transition @mariusandra I think it's reflecting what's happening pretty well. Insights _can be_ on dashboards, but they don't _belong_ to them. Especially since we plan to add support of a many-to-many relationship between these two models.

Twixes commented:
Examples:
![Zrzut ekranu 2021-11-9 o 16 08 56](https://user-images.githubusercontent.com/4550621/140952792-40f5849a-93f0-4d25-861d-2cdf64244904.png)
![Zrzut ekranu 2021-11-9 o 16 25 24](https://user-images.githubusercontent.com/4550621/140952904-1b0e0127-b0e8-47c6-8708-06a1ea2bc587.png)

Twixes committed:
Improve responsiveness more

mariusandra reviewed (approved):
This all looks good to me. I'd still find a way to get the true insight name inside the breadcrumbs when the URL refactor is done... but that's not today.

Other suggested changes:
1. Change "Session recordings" to "Recordings"
2. Change "Saved insights" to "Insights"

clarkus commented:
> Personally, I'm in the same boat as @paolodamico and it feels really weird that the breadcrumbs don't include the current URL. I believe it impedes usability... and this is also the approach recommended by my go to source for UX guidelines (NN Group). Point nr 4: Include the current page as the last item in the breadcrumb trail.

This feels more targeted to designs that separate page headers from breadcrumbs. We're not doing that. They're always shown together as a single header component. If they were shown separately, I'd be more inclined to repeat this, but based on our app hierarchy and the placement of this component I don't think we need to repeat the current page. Can you elaborate on the specific usability problems you're concerned about? 

> Added current location as the final breadcrumb. It shows the current item's name/key, except for the insight view, which doesn't have a solid way of getting the currently edited item from outside the page (fromItem crap which is due for being significantly refactored anyway for cleaner URLs).

I don't think we should do this. The header is meant to be an extension of the breadcrumbs. The header is a bold, prominent representation of the current location. I don't see what problem we're solving by repeating it in the breadcrumbs. 

> Removed underlines from links. Though now it's a bit less clear which breadcrumb items are links.

Shouldn't they all be links? Is this specific to showing the current item in the list of crumbs? If so, see my comment above. Not sure we should do this.

clarkus reviewed (changes requested):
Just making my comments about the current breadcrumb item an official request change for github.

mariusandra commented:
@clarkus It's hard to put a finger on it. Basically, for me the breadcrumbs are more or less useless without the final page in them, since I don't mentally register them as anything hierarchical without having an mental anchor to the current page in the list of links. I have no way of understanding that the current links are "everything up to the current page". I see the big title, and I know that's the title, but the font and style are so different, that the title and the crumbs could as well belong to a different and unconnected part of the page.

Without the current page in the breadcrumbs list, I'm personally 99% sure I'll use the sidebar or the back/forward buttons when I need to navigate, and the crumbs will just waste valuable vertical space.

That's my personal opinion, yet I'm quite sure the quote from the NNGroup site is for any kind of breadcrumbs, not for when they're followed by a big title.

mariusandra commented:
If we're worried about duplicating content... then, well, we're duplicating everything else in the breadcrumbs anyway, so why not make it complete?

clarkus commented:
> Without the current page in the breadcrumbs list, I'm personally 99% sure I'll use the sidebar or the back/forward buttons when I need to navigate, and the crumbs will just waste valuable vertical space.

The main purpose of the breadcrumbs is to give the user a sense of place - it's absolutely redundant with the left navigation, but that navigation is shown conditionally. It's meant to confirm that you're looking at the right thing when you have that nav collapsed. Yes it helps you get back quickly to the parent items, but you will need to use the navigation to get anywhere not in that hierarchy chain. 

> If we're worried about duplicating content... then, well, we're duplicating everything else in the breadcrumbs anyway, so why not make it complete?

Because this is repetition within the same component. You're literally saying `Organization > Project > Container > CurrentThing > CurrentThing`. My point is that the header is necessary to balance out placement of other page level actions, so taking that `CurrentThing` and giving it more prominence accomplishes the same task. The header is very much intended to be that current item. 

I drew the proposed breadcrumb solution to illustrate why I think it's a dead end. What does repeating `Insight #1234` accomplish over the header here? Is the heading using the same text value not achieving the same goal?

<img width="1431" alt="Screen Shot 2021-11-09 at 9 25 36 AM" src="https://user-images.githubusercontent.com/254612/140973983-c885c285-e575-4a13-8f61-46a111e3e22c.png">

Twixes commented:
I do agree with Chris's point about duplication (+ there's technical complexity in using internal data like insight name + inconsistencies in formatting can easily creep in e.g. "Events Stats" vs "Event stats"). Though I can see why some users (like Paolo and Marius) may have a better experience with that crumb. It seems that both sides feel pretty strongly about this so maybe we should just ask a couple more PostHog users to try this out and report feelings?

> Shouldn't they all be links?
@clarkus This is only loosely related to the final breadcrumb issue. The reason why in practice most of the breadcrumbs are not links, because the top-level ones (organization+project OR instance OR users) are not links. They are logical pointers, but not concrete ones. I _am_ a bit afraid that this may hinder usage of those breadcrumbs that _are_ links (because they point to concrete pages) if users are unsure what is and what isn't a link, but I have no great solution for that at the moment.

clarkus commented:
>  It seems that both sides feel pretty strongly about this so maybe we should just ask a couple more PostHog users to try this out and report feelings?

I am very happy and willing to change my stance if we have information around a problem this addresses. I just can't see the problem we're addressing. User's might have opinions, but I feel like it should be based on a clear problem first. 

> @clarkus This is only loosely related to the final breadcrumb issue. The reason why in practice most of the breadcrumbs are not links is because the top-level ones (organization + project OR instance OR user) aren't. They are logical pointers, but not concrete ones. I am a bit afraid that it may hinder usage of those breadcrumbs that are links (pointing to concrete pages) if users are unsure what is clickable, but I have no great solution for that at the moment.

I think it'd be fine to make those look look like unlinked text nodes in that case. We might eventually have more things built out around collaboration that gives us meaningful places to link to. Until then it's ok if they're not actionable. The core job is to show you where you are in the current context. If anything, an item looking like it isn't a link is a better UX because you know you can't click on the thing.

mariusandra commented:
Well, I still believe that **without** seeing a bold activated item inside the breadcrumbs, like here:

![image](https://user-images.githubusercontent.com/53387/141087881-6a27599b-ded2-4f4b-99c9-b6921b00e7c9.png)

... the breadcrumbs *lose all meaning* and just become a random selection of loosely disjoined links that are mostly not even links.

The header below the crumbs has a wildly different style, and you can't mentally link it into the breadcrumb navigation structure without [stopping to think](https://www.amazon.com/Dont-Make-Me-Think-Usability/dp/0321344758)... which equates with confusing UX.

That's also what the UX guidelines on the internet say, so ... 

I'm fine to go against common sense and remove them if this moves the project along, though it might be better to have a flag that enables/disables the last item, and ask for feedback from the rest of the company.

@clarkus I think the screenshot above could use a good bit of padding between the breadcrumbs and the header, then it won't look so cramped. More aggressively, I would even consider removing the header, as with it being duplicated in the breadcrumbs, we could save some vertical space.

Here are a few cases where having the breadcrumb looks perfectly fine to me (changing the last item style compared to what's in this PR):

![image](https://user-images.githubusercontent.com/53387/141090471-a3f2554f-415c-447d-8efe-02896a0d9b04.png)

![image](https://user-images.githubusercontent.com/53387/141090516-322229d8-0590-42bf-ac8f-9c03157ec60c.png)

![image](https://user-images.githubusercontent.com/53387/141090746-daeea0a0-ada6-4fe7-8a7c-55aeb0edd310.png)

![image](https://user-images.githubusercontent.com/53387/141090559-30f2f5ee-8be7-439a-9541-c4015ffcc816.png)

but of course if the header is so close to the crumbs it looks weird:

![image](https://user-images.githubusercontent.com/53387/141090665-98d53816-095d-4506-86ac-7a3ecccec889.png)

Twixes commented:
I've conducted qualitative research by means of messaging random PostHoggers who happened to be online on Slack. The subjects were presented with two demo GIFs showcasing a specific breadcrumbful user flow: one without that final breadcrumb, the other with it. The result is: **2 out of 2 subjects were strongly _in favor_ of the "You are here" breadcrumb**. Let's keep it then.

Twixes committed:
Merge branch 'master' into breadcrumbs

Twixes committed:
Add `breadcrumbsLogic` `path`

mariusandra commented:
The one change I'd make is to make the last part look like on the mocked screenshots I posted, and like figma: just black bold text instead of blue italic.

paolodamico commented:
@mariusandra, you've said it all (it's as if we read the same book), putting out a PR to propose those changes and address a few other bugs I noticed, #7025. Great approach too @Twixes!
